_src/_posts/2017-01-31-learning-rtb.markdown

```diff
@@ -0,0 +1,205 @@
+a succession of input events $(x_1,y_1), \ldots, (x_t,y_t)$. For this
+we minimize the sum over each previously seen input event of:
+(i) (The gradient of) the log-likelihood loss, which is a [logloss][] for the
+logistic model.
+(ii) A stabilizing term that introduces some strong convexity into the mix.
+(iii) Regularizing terms that combine lasso and ridge regularization.
+
+After filling in the details and translating everything to Greek:
+$$
+\theta_{t+1} = \underset{\theta}{\operatorname{argmin}}\ 
+{g_{1:t}}^T \theta +
+\frac{1}{2\eta_0}\sum_{s=1}^t ||\theta_s-\theta||^2_{A_s - A_{s-1}} +
+\lambda_1 ||\theta||_1 + \frac{\lambda_2}{2}||\theta||^2_2
+$$
+where:
+* $g_{1:t}$ is the sum of the previous gradients $g_1, \ldots, g_t$ of the
+  loss function.
+* $A_t = (\sum_{s=1}^t g_s g'_s)^{1/2}$, which
+  [minimizes the regret bound][minregret] over Mahalanobis norms for projected
+  gradient descent. That said, in real life we just consider the diagonal matrix
+  $diag(A_t)$ for the sake of computation.
+* $\eta_0$ is the initial learning step.
+* $\lambda_1$ is the strength of the lasso regularization.
+* $\lambda_2$ is the strength of the ridge regularization.
+
+Regularization terms aside, this is mostly equivalent to an online gradient
+descent formulation with learning rate $\eta_t = \eta_0 A_t^{-1}$, but solving
+the minimization in a follow-the-leader fashion allows for an effective
+implementation of lasso regularization that produces coefficients which are
+exactly zero (and, hence, more sparse models that are cheaper to store and
+transfer and, most importantly, way faster to evaluate). Although at first sight
+it might seem that implementing the follow-the-leader update step ---which sums
+over every previous input--- would be harder than implementing the gradient
+descent update step ---which just cares about the last update--- this first
+impression turns out to be wrong after a [careful reformulation][trenches] of
+the above expression:
+$$
+\theta_{t+1} = \underset{\theta}{\operatorname{argmin}}\ 
+{z_t}^T \cdot \theta +
+\frac{1}{2\eta_0} ||\theta||^2_{A_t} +
+\lambda_1 ||\theta||_1 + \frac{\lambda_2}{2}||\theta||^2_2
+$$
+where $z_t = {g_{1:t}} - \frac{1}{\eta_0} \sum_{s=1}^t (A_s - A_{s-1})\theta_s$,
+which can be cheaply calculated in an incremental way as:
+$$
+z_t = z_{t-1} + g_t + \frac{1}{\eta_0} (A_t - A_{t-1})\theta_t
+$$
+Now, when $A_t$ is diagonal the closed form of the solution is just:
+$$
+\theta_{t,i} =
+\begin{cases}
+0 & \text{if } |z_{t,i}| \le \lambda_1 \\
+-\frac{z_{t,i} - \text{sgn}(z_{t,i})\lambda_1}{A_{t,i}/\eta_0 - \lambda_2} & \text{otherwise}
+\end{cases}
+$$
+
+Glossing over many technicalities like:
+* Keeping a huge inverse hash map, in order to recover features from hashes.
+* Following a "memory schedule" that weights each input according to the time
+  elapsed since its arrival, in order to adapt to ever-changing market
+  conditions[^decay].
```

Comment:
As above, usually lists have ";", more even so here that they are part of a bigger sentence ("... our ").